Cython/Compiler/ExprNodes.py

```diff
@@ -12414,7 +12419,14 @@ def infer_builtin_types_operation(self, type1, type2):
 
     def analyse_operation(self, env):
         self._check_truedivision(env)
-        NumBinopNode.analyse_operation(self, env)
+        result = NumBinopNode.analyse_operation(self, env)
+
+        # The assumption here is that result is either 'self' or a coercion
+        # node containing 'self'. Thus it is reasonable to keep manipulating
+        # 'self' even if it's been replaced as the eventual result.
+        from . import Visitor
+        assert Visitor.tree_contains(result, self)
```

Comment:
That's a bit of an expensive assertion. It's not very likely to strike at some point, and given the potential impact of not doing any of the operations below, we'd probably notice it anyway if we make an unanticipated change in the base class.

I'd rather remove the assertion and instead add a comment to the `return node` in the base class implementation that this subclass expects to be able to modify the original node after the fact. That should catch the most obvious risk of breakage. The same applies to the upcall to `DivNode` further down.

With that change, I'd be happy to see this merged.